modules/dmrpp_module/build_dmrpp_util.cc

```diff
@@ -0,0 +1,963 @@
+        if (false == is_scalar)
+            H5Sclose(mspace);
+        H5Tclose(memtype);
+        H5Tclose(dtypeid);
+        H5Sclose(dspace);
+        throw InternalErr (__FILE__, __LINE__, "Fail to read the HDF5 variable length string dataset.");
+    }
+
+    // For scalar, nelms is 1.
+    char*temp_bp = &strval[0];
+    char*onestring = NULL;
+    for (int i =0;i<nelms;i++) {
+        onestring = *(char**)temp_bp;
+        if(onestring!=NULL )
+            finstrval[i] =string(onestring);
+        else // We will add a NULL if onestring is NULL.
+            finstrval[i]="";
+        temp_bp +=ty_size;
+    }
+
+    if (false == strval.empty()) {
+        herr_t ret_vlen_claim;
+        if (true == is_scalar)
+            ret_vlen_claim = H5Dvlen_reclaim(memtype,dspace,H5P_DEFAULT,(void*)&strval[0]);
+        else
+            ret_vlen_claim = H5Dvlen_reclaim(memtype,mspace,H5P_DEFAULT,(void*)&strval[0]);
+        if (ret_vlen_claim < 0){
+            if (false == is_scalar)
+                H5Sclose(mspace);
+            H5Tclose(memtype);
+            H5Tclose(dtypeid);
+            H5Sclose(dspace);
+            throw InternalErr (__FILE__, __LINE__, "Cannot reclaim the memory buffer of the HDF5 variable length string.");
+
+        }
+    }
+
+    if (false == is_scalar)
+        H5Sclose(mspace);
+    H5Tclose(memtype);
+    H5Tclose(dtypeid);
+    H5Sclose(dspace);
+
+    return true;
+
+}
+
+/**
+ * @brief Print information about the data type
+ *
+ * @note Calling this indicates that the build_dmrpp utility could not
+ * get chunk information, and is probably an error, but I'm not sure
+ * that's always the case. jhrg 5/7/18
+ *
+ * @param dataset
+ * @param layout_type
+ */
+static void print_dataset_type_info(hid_t dataset, uint8_t layout_type) {
+    hid_t dtype_id = H5Dget_type(dataset);
+    if (dtype_id < 0) {
+        throw BESInternalError("Cannot obtain the correct HDF5 datatype.", __FILE__, __LINE__);
+    }
+
+    if (H5Tget_class(dtype_id) == H5T_INTEGER || H5Tget_class(dtype_id) == H5T_FLOAT) {
+        hid_t dcpl_id = H5Dget_create_plist(dataset);
+        if (dcpl_id < 0) {
+            throw BESInternalError("Cannot obtain the HDF5 dataset creation property list.", __FILE__, __LINE__);
+        }
+
+        try {
+            // Wrap the resources like dcpl_id in try/catch blocks so that the
+            // calls to H5Pclose(dcpl_id) for each error can be removed. jhrg 5/7/18
+            H5D_fill_value_t fvalue_status;
+            if (H5Pfill_value_defined(dcpl_id, &fvalue_status) < 0) {
+                H5Pclose(dcpl_id);
+                throw BESInternalError("Cannot obtain the fill value status.", __FILE__, __LINE__);
+            }
+            if (fvalue_status == H5D_FILL_VALUE_UNDEFINED) {
+                // Replace with switch(), here and elsewhere. jhrg 5/7/18
+                if (layout_type == 1)
+                    cerr << " The storage size is 0 and the storage type is contiguous." << endl;
```

Comment:
If this file is commented out. Please ignore my review. Otherwise, I don't understand your logic here. I think you copy the code from an example code I gave to you but didn't do in the correct context. If the fillvalue is not defined, it doesn't mean the storage size is 0.  Maybe this is  only for the debugging purpose. If this is  the case, you may mark it differently.